### PR TITLE
Add alerts for zync and zync-database availability 

### PR DESCRIPTION
**Pull request by austincunningham**
# Why
*JIRA:* https://issues.redhat.com/browse/INTLY-5290

# Verification
- checkout this branch
- run rhmi locally 
- check the application monitoring prometheus alerts for the following alerts
ThreeScaleZyncPodAvailiblity, ThreeScaleZyncDatabasePodAvailiblity
image

- Test the alerts going to the deployment configs and scaling down the zync and zync-database pods for 5 minutes to confirm the alert triggers
image
>*NOTE*: zync pods may scale up again so may need to stop the 3scale operator.

### CODE REVIEW

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml [hunk 1]
```diff
@@ -94,6 +94,24 @@ spec:
         for: 5m
         labels:
           severity: critical
+      - alert: ThreeScaleZyncPodAvailiblity
```
Comment (david-martin) [inline templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: minor, but spelling of availability

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml [hunk 2]
```diff
@@ -94,6 +94,24 @@ spec:
...
+          absent(count((kube_pod_status_ready{condition="true",namespace="redhat-rhmi-3scale",pod=~"zync-1.*"}) + on(pod) (kube_pod_container_status_running{namespace="redhat-rhmi-3scale",container="zync"}))==2)
+        for: 5m
+        labels:
+          severity: critical
+      - alert: ThreeScaleZyncDatabasePodAvailiblity
+        annotations:
+          message:  3Scale Zync-database has no pods in a ready state.
+          sop_url: https://github.com/RHCloudServices/integreatly-help/blob/master/sops/alerts_and_troubleshooting.md
+        expr: |
+          absent(sum(kube_pod_container_status_running{namespace="redhat-rhmi-3scale",pod=~"zync-database.*", container="postgresql"})==1)
```
Comment (david-martin) [inline templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: @austincunningham would it make sense to check the pod status rather than container?
The pod status is what determines if it's available to use (which in turn ensures all containers are ready)

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml (hunk 1 above)
Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: :+1:

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml (hunk 2 above)
Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: Will have a look :eyes:  in prometheus to see if it will return what we want.

Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: Think the reason we used kube_pod_container_status_running was it didn't return the deployment pod e.g.
kube_pod_status_ready{ namespace="redhat-rhmi-3scale", pod=~"zync-database.*" }
```
kube_pod_status_ready{condition="true",endpoint="https-main",instance="10.131.2.5:8443",job="kube-state-metrics",namespace="redhat-rhmi-3scale",pod="zync-database-1-deploy",prometheus="openshift-monitoring/k8s",service="kube-state-metrics"} | 0
-- | --
kube_pod_status_ready{condition="true",endpoint="https-main",instance="10.131.2.5:8443",job="kube-state-metrics",namespace="redhat-rhmi-3scale",pod="zync-database-1-rtqrt",prometheus="openshift-monitoring/k8s",service="kube-state-metrics"}
```

Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: @david-martin will use
```
absent(sum(kube_pod_status_ready{ namespace="redhat-rhmi-3scale", pod=~"zync-database.*",condition="true" })==1)
```

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml [hunk 3]
```diff
@@ -94,6 +94,24 @@ spec:
         for: 5m
         labels:
           severity: critical
+      - alert: ThreeScaleZyncPodAvailability
+        annotations:
+          message:  3Scale Zync has no pods in a ready state.
+          sop_url: https://github.com/RHCloudServices/integreatly-help/blob/master/sops/alerts_and_troubleshooting.md
+        expr: |
+          absent(count((kube_pod_status_ready{condition="true",namespace="redhat-rhmi-3scale",pod=~"zync-1.*"}) + on(pod) (kube_pod_container_status_running{namespace="redhat-rhmi-3scale",container="zync"}))==2)
```
Comment (david-martin) [inline templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: as discussed in person, this may start triggering if a new deploy is kicked off (incrementing the deploy number).
I'd suggest matching based on pod labels (by joining with kube_pod_labels_info metric)

Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: agree, changed the rule to check for label
```
((kube_pod_labels{ namespace="redhat-rhmi-3scale", pod=~"zync-database.*",label_threescale_component_element="database"})+on(pod)(kube_pod_status_ready{ namespace="redhat-rhmi-3scale", pod=~"zync-database.*",condition="true" })!=2)```

[the remark this answers is not in the document]
Comment (david-martin) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: good catch @obrienrobert .
@austincunningham Does that make sense?

[the remark this answers is not in the document]
Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: :+1: good suggestion.

**Comment by obrienrobert** [comment]
@austincunningham as I mentioned, I scaled the `zync-db` pods down  from 2 to 0. The alert `ThreeScaleZyncDatabasePodAvailiblity` hasn't triggered (stayed green).

**Comment by obrienrobert** [comment]
Verified that both alerts work as expected using the verification steps.

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml [hunk 4]
```diff
@@ -94,6 +94,24 @@ spec:
...
+          absent(count((kube_pod_status_ready{condition="true",namespace="{{ index .Params "Namespace" }}",pod=~"zync-1.*"}) + on(pod) (kube_pod_container_status_running{namespace="{{ index .Params "Namespace" }}",container="zync"}))==2)
+        for: 5m
+        labels:
+          severity: critical
+      - alert: ThreeScaleZyncDatabasePodAvailiblity
+        annotations:
+          message:  3Scale Zync-database has no pods in a ready state.
+          sop_url: https://github.com/RHCloudServices/integreatly-help/blob/master/sops/alerts_and_troubleshooting.md
+        expr: |
+          absent(count((kube_pod_labels{ namespace="{{ index .Params "Namespace" }}", pod=~"zync-database.*",label_threescale_component_element="database"}) + on(pod) (kube_pod_status_ready{ namespace="{{ index .Params "Namespace" }}", pod=~"zync-database.*",condition="true" })==2))
```
Comment (david-martin) [inline templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: Can remove the pod regex and just use labels.

i.e. 
```
absent(count(kube_pod_status_ready{condition="true",namespace="redhat-rhmi-3scale"} * on (pod, namespace) kube_pod_labels{namespace="redhat-rhmi-3scale", label_threescale_component="zync", label_threescale_component_element="database"}) == 1)
```

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml [hunk 5]
```diff
@@ -94,6 +94,24 @@ spec:
         for: 5m
         labels:
           severity: critical
+      - alert: ThreeScaleZyncPodAvailability
+        annotations:
+          message:  3Scale Zync has no pods in a ready state.
+          sop_url: https://github.com/RHCloudServices/integreatly-help/blob/master/sops/alerts_and_troubleshooting.md
+        expr: |
+          absent(count((kube_pod_status_ready{condition="true",namespace="{{ index .Params "Namespace" }}",pod=~"zync-1.*"}) + on(pod) (kube_pod_container_status_running{namespace="{{ index .Params "Namespace" }}",container="zync"}))==2)
```
Comment (david-martin) [inline templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: Can remove the pod regex and just use labels.

i.e.


```
absent(count(kube_pod_status_ready{condition="true",namespace="redhat-rhmi-3scale"} * on (pod, namespace) kube_pod_labels{namespace="redhat-rhmi-3scale", label_threescale_component="zync", label_threescale_component_element=""}) == 2)
```

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml [hunk 6]
```diff
@@ -94,6 +94,24 @@ spec:
         for: 5m
         labels:
           severity: critical
+      - alert: ThreeScaleZyncPodAvailability
+        annotations:
+          message:  3Scale Zync has no pods in a ready state.
```
Comment (david-martin) [inline templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: This message is not completely accurate. The alert will trigger if there not exactly 2 zync pods in a ready state.

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml (hunk 4 above)
Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: :+1:

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml (hunk 5 above)
Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: :+1:

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml (hunk 6 above)
Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: :+1:

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml [hunk 7]
```diff
@@ -94,15 +94,24 @@ spec:
         for: 5m
         labels:
           severity: critical
-      - alert: ThreeScalePodCount
```
Comment (david-martin) [inline templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: 👍  Makes sense to remove this alert as there's a lot more specific alerts for the various components of 3scale

**Comment by david-martin** [comment]
/approve

Would like to get an lgtm from @obrienrobert  as well

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml [hunk 8]
```diff
@@ -94,15 +94,24 @@ spec:
...
+      - alert: ThreeScaleZyncPodAvailability
         annotations:
+          message:  3Scale Zync has {{ "{{" }} printf "%.0f" $value {{ "}}" }} in a ready state. Expected at least 2 pod.
```
Comment (obrienrobert) [inline templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: Should `Expected at least 2 pod` be `Expected at least 2 pods` (plural)? 

Or to be more descriptive, `Expected a minimum of 2 pods`. Just a suggestion.

Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: :+1:

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml (hunk 7 above)
Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: agree

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml [hunk 9]
```diff
@@ -94,15 +94,24 @@ spec:
...
           sop_url: https://github.com/RHCloudServices/integreatly-help/blob/master/sops/alerts_and_troubleshooting.md
-          message: Pod count for namespace {{ "{{" }} $labels.namespace {{ "}}" }} is {{ "{{" }} printf "%.0f" $value {{ "}}" }}. Expected at least 15 pods.
         expr: |
-          absent(sum(kube_pod_status_ready{condition="true",namespace="{{ index .Params "Namespace" }}"}) >= 14)
+          (1-absent(kube_pod_status_ready{condition="true",namespace="{{ index .Params "Namespace" }}"} * on (pod, namespace) kube_pod_labels{namespace="{{ index .Params "Namespace" }}", label_threescale_component="zync", label_threescale_component_element=""})) or count(kube_pod_status_ready{condition="true",namespace="{{ index .Params "Namespace" }}"} * on (pod, namespace) kube_pod_labels{namespace="{{ index .Params "Namespace" }}", label_threescale_component="zync", label_threescale_component_element=""}) != 2
         for: 5m
         labels:
-          severity: warning
+          severity: critical
+      - alert: ThreeScaleZyncDatabasePodAvailiblity
```
Comment (obrienrobert) [inline templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: Just noticed typo in `Availiblity`-> `Availability`

Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: :+1:

File: templates/monitoring/kube_state_metrics_3scale_alerts.yaml [hunk 10]
```diff
@@ -94,15 +94,24 @@ spec:
         for: 5m
         labels:
           severity: critical
-      - alert: ThreeScalePodCount
+      - alert: ThreeScaleZyncPodAvailability
         annotations:
+          message:  3Scale Zync has {{ "{{" }} printf "%.0f" $value {{ "}}" }} pods in a ready state. Expected at minimum of 2 pods.
```
Comment (obrienrobert) [inline templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: One small thing - `Expected at minimum of 2 pods`. I would reword this to: `Expected a minimum of 2 pods`. Same for the other alert.

Comment (austincunningham) [reply templates/monitoring/kube_state_metrics_3scale_alerts.yaml]: thanks @obrienrobert  :+1:
